src/main/scala/Validator.scala

```diff
@@ -1,115 +1,387 @@
+          mapAvailableCharsets(dialect.encoding),
+          format
+        )
+      )
     } else {
-      CSVParser.parse(
-        tableUri.toURL,
-        StandardCharsets.UTF_8,
-        CSVFormat.DEFAULT
+      Right(
+        CSVParser.parse(
+          tableUri.toURL,
+          mapAvailableCharsets(dialect.encoding),
+          format
+        )
       )
     }
+  }
+
+  def readAndValidateCsv(
+      schema: TableGroup,
+      tableUri: URI,
+      format: CSVFormat,
+      dialect: Dialect
+  ): (
+      WarningsAndErrors,
+      Map[ChildTableForeignKey, Set[KeyWithContext]],
+      Map[ParentTableForeignKeyReference, Set[KeyWithContext]]
+  ) = {
+    getParser(tableUri, dialect, format) match {
+      case Right(parser) =>
+        readAndValidateWithParser(schema, tableUri, dialect, parser)
+      case Left(error) => {
+        val errors = Array(error)
+        (WarningsAndErrors(Array(), errors), Map(), Map())
+      }
+    }
+
+  }
+
+  private def readAndValidateWithParser(
+      schema: TableGroup,
+      tableUri: URI,
+      dialect: Dialect,
+      parser: CSVParser
+  ) = {
+    var warnings: Array[WarningWithCsvContext] = Array()
+    var errors: Array[ErrorWithCsvContext] = Array()
+
+    val parserAfterSkippedRows = parser.asScala.drop(dialect.skipRows)
     val table = schema.tables(tableUri.toString)
+    val childTableForeignKeys =
+      Map[ChildTableForeignKey, mutable.Set[KeyWithContext]]()
+    val parentTableForeignKeyReferences =
+      Map[ParentTableForeignKeyReference, mutable.Set[KeyWithContext]]()
+    val allPrimaryKeyValues: mutable.Set[List[Any]] =
+      Set() // List of type Any with same values are not added again in a Set, whereas Array of Type any behaves
+    // differently.
 
-    // List of type Any with same values are not added again in a Set, whereas Array of Type any behaves differently.
-    var allPrimaryKeyValues: Set[List[Any]] = Set()
-    for (row <- parser.asScala) {
-      if (row.getRecordNumber == 1 && csvHeaderExpected) {
-        validateHeader(schema, row, tableUri)
+    parserAfterSkippedRows
+      .map(row =>
+        (
+          row,
+          parseRow(
+            schema,
+            tableUri,
+            dialect,
+            table,
+            row
+          )
+        )
+      )
+      .foreach {
+        case (row, validateRowOutput) => {
+          errors ++= validateRowOutput.warningsAndErrors.errors
+          warnings ++= validateRowOutput.warningsAndErrors.warnings
+          setChildTableForeignKeys(
+            validateRowOutput,
+            childTableForeignKeys
+          )
+          setParentTableForeignKeyReferences(
+            validateRowOutput,
+            parentTableForeignKeyReferences
+          )
+          validatePrimaryKey(allPrimaryKeyValues, row, validateRowOutput)
+            .ifDefined(e => {
```

Comment:
May as well turn this back into a one liner

Reply:
Not exactly sure what you meant by this. If this is what you want

`validatePrimaryKey(allPrimaryKeyValues, row, validateRowOutput).ifDefined(e => { errors :+= e })` 

scala fmt is not allowing me to do so

Reply:
`.ifDefined(e => errors :+= e)`

Reply:
Oh okay, got it. Changed